src/factory-reset/factory-reset-tool.c

```diff
@@ -0,0 +1,381 @@
+
+        r = sd_device_enumerator_allow_uninitialized(e);
+        if (r < 0)
+                return log_error_errno(r, "Failed to enable enumeration of uninitialized devices: %m");
+
+        r = sd_device_enumerator_add_match_subsystem(e, "block", /* match = */ true);
+        if (r < 0)
+                return log_error_errno(r, "Failed to filter device enumeration by 'block' subsystem: %m");
+
+        if (!arg_quiet)
+                log_info("Retriggering block devices.");
+
+        FOREACH_DEVICE(e, d) {
+                r = sd_device_trigger(d, SD_DEVICE_CHANGE);
+                if (r < 0)
+                        log_device_debug_errno(d, r, "Failed to trigger block device, ignoring: %m");
+        }
+
+        return 0;
+}
+
+static int verb_complete(int argc, char *argv[], void *userdata) {
+        int r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return log_error_errno(f, "Failed to dermine factory reset state: %m");
+        log_debug("Current factory state is: %s", factory_reset_mode_to_string(f));
+        if (f != FACTORY_RESET_ON) {
+                log_info("Attempted to leave factory reset mode, even though we are not in factory reset mode. Ignoring.");
+                return 0;
+        }
+
+        r = efi_set_variable(EFI_SYSTEMD_VARIABLE_STR("FactoryResetRequest"), /* value= */ NULL, /* size= */ 0);
+        if (r < 0)
+                log_full_errno(r == -ENOENT ? LOG_DEBUG : LOG_WARNING, r,
+                               "Failed to remove FactoryResetRequest EFI variable: %m");
+
+        r = touch("/run/systemd/factory-reset-complete");
+        if (r < 0)
+                return log_error_errno(r, "Failed to create /run/systemd/factory-reset-complete file: %m");
+
+        if (!arg_quiet)
+                log_info("Successfully left factory reset mode.");
+
+        if (arg_retrigger)
+                (void) retrigger_block_devices();
+
+        return 0;
+}
+
+static int vl_method_get_factory_reset_mode(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return f;
+
+        return sd_varlink_replybo(link, SD_JSON_BUILD_PAIR_STRING("mode", factory_reset_mode_to_string(f)));
+}
+
+static int vl_method_can_request_factory_reset(sd_varlink *link, sd_json_variant *parameters, sd_varlink_method_flags_t flags, void *userdata) {
+        int r;
+
+        assert(parameters);
+
+        r = sd_varlink_dispatch(link, parameters, /* table= */ NULL, /* userdata= */ NULL);
+        if (r != 0)
+                return r;
+
+        FactoryResetMode f = factory_reset_mode();
+        if (f < 0)
+                return f;
```

Comment:
Missing mode check? That is, refuse when `f` is ON or PENDING ?

Reply:
nope, just copy pasta, will remove